spinalcordtoolbox/scripts/sct_analyze_lesion.py

```diff
@@ -584,70 +595,91 @@ def __relative_ROIvol_in_mask(self, im_mask_data, im_atlas_roi_data, p_lst, im_t
-                im_vert_cur[np.where(im_vert_cur != vert)] = 0
-            else:
-                im_vert_cur = None
-            if im_vert_cur is None or np.count_nonzero(im_vert_cur * np.copy(im_lesion)):
+        im_vert_and_lesion = im_vert * im_lesion  # to check which vertebral levels have lesions
+        # loop over slices/vertlevels
+        for row, indices_to_keep in sct_progress_bar(rows_with_total.items(), unit=self.row_name,
+                                                     desc="  Computing ROI distribution for all lesions"):
+            if row == total_row or np.count_nonzero(im_vert_and_lesion[indices_to_keep]):
                 res_perTract_dct = {}  # for each tract compute the volume occupied by lesion and the volume of the tract
-                idx = self.distrib_matrix_dct[sheet_name][self.distrib_matrix_dct[sheet_name].vert == str(vert)].index
+                idx = self.distrib_matrix_dct[sheet_name][self.distrib_matrix_dct[sheet_name].row == str(row)].index
                 for tract_id in atlas_data:  # loop over the tracts
-                    res_perTract_dct[tract_id] = self.__relative_ROIvol_in_mask(im_mask_data=np.copy(im_lesion),
-                                                                                im_atlas_roi_data=np.copy(atlas_data[tract_id]),
+                    res_perTract_dct[tract_id] = self.__relative_ROIvol_in_mask(im_mask_data=im_lesion,
+                                                                                im_atlas_roi_data=atlas_data[tract_id],
                                                                                 p_lst=p_lst,
-                                                                                im_template_vert_data=np.copy(im_vert_cur),
-                                                                                vert_level=vert)
+                                                                                indices_to_keep=indices_to_keep)
 
-                # group tracts to compute involvement in GM, WM, DC, VF, LF
-                self.distrib_matrix_dct[sheet_name].loc[idx, 'PAM50_GM'] = self.__regroup_per_tracts(vol_dct=res_perTract_dct, tract_limit=[30, 35])
-                self.distrib_matrix_dct[sheet_name].loc[idx, 'PAM50_WM'] = self.__regroup_per_tracts(vol_dct=res_perTract_dct, tract_limit=[0, 29])
-                self.distrib_matrix_dct[sheet_name].loc[idx, 'PAM50_DC'] = self.__regroup_per_tracts(vol_dct=res_perTract_dct, tract_limit=[0, 3])
-                self.distrib_matrix_dct[sheet_name].loc[idx, 'PAM50_VF'] = self.__regroup_per_tracts(vol_dct=res_perTract_dct, tract_limit=[14, 29])
-                self.distrib_matrix_dct[sheet_name].loc[idx, 'PAM50_LF'] = self.__regroup_per_tracts(vol_dct=res_perTract_dct, tract_limit=[4, 13])
+                # group tracts to compute involvement in CombinedLabels (GM, WM, DC, VF, LF)
+                for label_name, sublabels in self.atlas_combinedlabels.items():
```

Comment:
This will add a 6th one for the whole spinal cord, and change the label names, presumably that's ok?

From `data/PAM50/atlas/info_label.txt`:
```
# Combined labels
# Keyword=CombinedLabels (Please DO NOT change this line)
# ID, name, IDgroup
50, spinal cord, 0:35
51, white matter, 0:29
52, gray matter, 30:35
53, dorsal columns, 0:3
54, lateral funiculi, 4:13
55, ventral funiculi, 14:29
```

Reply:
This is a really good point, and something I had been thinking about too, but for a different reason: The "whole spinal cord" is identical to the existing "total" column, meaning that the column is duplicated.

This could be a reason to ignore the spinal cord column, but it also feels like a good sanity check (if the columns don't match, something has gone wrong). It also kind of clarifies the meaning of the "total" column a bit: the "total" is equivalent to the lesion distribution across all tracts i.e. the full spinal cord.

Reply:
I think https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4584/commits/89881dccc0e91c11b3257d36904bb524fa80c847 might be a good compromise? In that, we exclude `spinal cord` but _only_ if it will match the total column.

If anything doesn't match (for example if the `info_label.txt` file is modified to include the CSF file for some reason), then the `spinal cord` label will be preserved.

Reply:
That's an elegant solution, I like it!